lib/schedule/data.ex

```diff
@@ -66,34 +80,75 @@ defmodule Schedule.Data do
+    {:swings, [:service_id_and_route_id, :swings], []}
+  ]
+
+  @spec all_routes(tables()) :: [Route.t()]
+  def all_routes(%{routes: routes_table}) do
+    :mnesia.ets(fn ->
+      routes_table
+      |> :ets.tab2list()
+      |> Enum.map(fn {_route_id, route} -> route end)
+    end)
+  end
 
-  @spec timepoints_on_route(t(), Route.id()) :: [Timepoint.t()]
-  def timepoints_on_route(
-        %__MODULE__{timepoints_by_route: timepoints_by_route},
-        route_id
-      ),
-      do: Map.get(timepoints_by_route, route_id, [])
+  @spec timepoints_on_route(tables(), Route.id()) :: [Timepoint.t()]
+  def timepoints_on_route(%{timepoints_on_route: timepoints_on_route_table}, route_id) do
+    case :mnesia.dirty_read(timepoints_on_route_table, route_id) do
+      [{^timepoints_on_route_table, _, timepoints}] -> timepoints
+      _ -> []
+    end
+  end
 
-  @spec timepoint_names_by_id(t()) :: Timepoint.timepoint_names_by_id()
-  def timepoint_names_by_id(%__MODULE__{timepoint_names_by_id: timepoint_names_by_id}),
-    do: timepoint_names_by_id
+  @spec timepoint_names_by_id(tables()) :: Timepoint.timepoint_names_by_id()
+  def timepoint_names_by_id(%{timepoint_names_by_id: timepoint_names_by_id_table}) do
+    :mnesia.ets(fn ->
+      timepoint_names_by_id_table
+      |> :ets.tab2list()
+      |> Map.new()
+    end)
+  end
 
-  @spec stop(t(), Stop.id()) :: Stop.t() | nil
-  def stop(%__MODULE__{stops: stops}, stop_id), do: stops[stop_id]
+  @spec stop(tables(), Stop.id()) :: Stop.t() | nil
+  def stop(%{stops: stops_table}, stop_id) do
+    case :mnesia.dirty_read(stops_table, stop_id) do
+      [{^stops_table, _, stop}] -> stop
+      _ -> nil
+    end
+  end
 
-  @spec trip(t(), Schedule.Trip.id()) :: Schedule.Trip.t() | nil
-  def trip(%__MODULE__{trips: trips}, trip_id), do: trips[trip_id]
+  @spec trip(tables(), Schedule.Trip.id()) :: Schedule.Trip.t() | nil
+  def trip(%{trips: trips_table}, trip_id) do
+    case :mnesia.dirty_read(trips_table, trip_id) do
+      [{^trips_table, _, trip}] -> trip
+      _ -> nil
+    end
+  end
 
-  @spec trips_by_id(t(), [Schedule.Trip.id()]) :: %{Schedule.Trip.id() => Schedule.Trip.t()}
-  def trips_by_id(%__MODULE__{trips: trips}, trip_ids) do
-    Map.take(trips, trip_ids)
+  # TODO: can this be done in a more performant way?
+  @spec trips_by_id(tables(), [Schedule.Trip.id()]) :: %{Schedule.Trip.id() => Schedule.Trip.t()}
+  def trips_by_id(tables, trip_ids) do
+    trip_ids
+    |> Enum.map(&trip(tables, &1))
+    |> Enum.filter(&(!is_nil(&1)))
   end
 
-  @spec block(t(), Block.id(), Service.id()) :: Block.t() | nil
-  def block(%__MODULE__{blocks: blocks}, schedule_id, block_id) do
-    Block.get(blocks, schedule_id, block_id)
+  @spec block(tables(), Block.id(), Service.id()) :: Block.t() | nil
+  def block(%{blocks: blocks_table}, schedule_id, block_id) do
+    {blocks_table, {block_id, schedule_id}, :_, :_}
+    |> :mnesia.dirty_match_object()
+    |> Enum.map(&elem(&1, 3))
```

Comment:
suggestion: you should also be able to use `:mnesia.dirty_select` to only pull out the data you need.

Reply:
Also, this is returning a list but the spec says it returns a `Block` or `nil`.